Fix #228 - Add organizers management views to admin

olasitarska opened the pull request:
This pull request makes following changes:

- Adds a manage organizers view in admin, that allows superusers and organizers to remove organizers from upcoming events. 
- Adds an "add organizers" view in admin, that allows superuser and organizers to add organizers to upcoming events.
- Refactors "add_organizer" and "new_event" commands to use one form. Moves invitation to Slack into the form too.
- Adds sending automatic emails whenever organizer is added to event. This means we no longer need to send organizers passwords in emails manually. 
- Organizers can only add/remove organizers from their upcoming events. Superusers can do it for all upcoming events. 

@aniav your feedback on technical things here would be appreciated! I saw you recently refactored some of the commands too, so I wanna make sure we're consistent ✨ 
@patjouk your feedback on the functionality and whether it helps would be super helpful too!

olasitarska commented:
This is for issue #228

aniav commented on core/admin.py:
```diff
@@ -1,16 +1,20 @@
 from django.contrib.flatpages.models import FlatPage
 from django.contrib.flatpages.admin import FlatPageAdmin, FlatpageForm
 from django.utils.safestring import mark_safe
+from django.http import Http404
```
Unused import 😜

aniav commented on core/admin.py:
```diff
@@ -44,6 +48,84 @@ def get_readonly_fields(self, request, obj=None):
+            user = User.objects.get(id=request.GET['remove'])
+            if user in event.team.all():
+                event.team.remove(user)
+                event.save()
```
You don't need to save the event object to have it's m2m relationships updated.

aniav commented on core/forms.py:
```diff
@@ -14,6 +21,87 @@ def clean(self, values):
+        if event_choices:
+            self.fields['event'].queryset = event_choices
+
+    def generate_password(self):
```
You can use [make_random_password](https://docs.djangoproject.com/en/1.9/topics/auth/customizing/#django.contrib.auth.models.BaseUserManager.make_random_password) instead 🙂

aniav commented on core/forms.py:
```diff
@@ -14,6 +21,87 @@ def clean(self, values):
+            'user': user,
+            'event': self.cleaned_data['event']
+        })
+        subject = 'You have been granted to new Django Girls event'
```
It should be "granted access" or "added" 😇

aniav commented on core/forms.py:
```diff
@@ -14,6 +21,87 @@ def clean(self, values):
+            first_name = self.cleaned_data['name'].split(' ')[0]
+            last_name = self.cleaned_data['name'].replace(first_name, '')
+            user = User.objects.create(email=email,
+                                        first_name=first_name,
```
Nitpick: the indentation doesn't match the previous line 😇

aniav commented on core/forms.py:
```diff
@@ -14,6 +21,87 @@ def clean(self, values):
+            user.save()
+            user.groups.add(1)
+            event.team.add(user)
+            event.save()
```
Sam as earlier, no need to save the event object :)

aniav commented on core/forms.py:
```diff
@@ -14,6 +21,87 @@ def clean(self, values):
+        self._errors = []
+        email = self.cleaned_data['email']
+        event = self.cleaned_data['event']
+        if not User.objects.filter(email=email).exists():
```
Ideally you could use `User.objects.get_or_create` to avoid multiple queries to the User table 😀

aniav commented on templates/emails/setup.txt:
```diff
@@ -21,10 +21,8 @@ Or you can forward incoming mail to your private email account by following thes
-http://djangogirls.org/admin/
-{% for member in members %}{{ member.first_name }} - email: {{ member.email }} password: {{ member.password }}
-{% endfor %}
+You should just receive a separate email with auto generated password and instructions to login. 
```
Should be: "You should receive (...) in a moment" or "We have just sent you an e-mail" or "You should have just received" :)

olasitarska commented:
Oh I totally missed these comments! Will fix it all today 🔨

patjouk commented on core/forms.py:
```diff
@@ -14,6 +18,78 @@ def clean(self, values):
+    (including password).
+    """
+    event = forms.ModelChoiceField(queryset=Event.objects.all())
+    name = forms.CharField(label="Organizer's full name")
```
I would change it for "Organizer's first and last name". That way, we will have the full name (some organizers only send the first name when I ask for full name) and it will be in the order we need.

patjouk commented on core/admin.py:
```diff
@@ -44,6 +49,83 @@ def get_readonly_fields(self, request, obj=None):
+                user = form.save()
+                messages.success(request,
+                    '{} has been added to your event, yay!'.format(user.get_full_name()))
+                return redirect('/admin/core/event/manage_organizers/?event_id={}'.format(event.id))
```
It may be less confusing to return to the "add organizers page". They usually ask me to add a bunch of people at the same time.
There is also a problem with the redirection: if you have more than one event, it will redirect you to the first event in alphabetical order.

patjouk commented on templates/admin/core/event/view_manage_organizers.html:
```diff
@@ -0,0 +1,38 @@
+{% extends "admin/base_site.html" %}
+
+{% block content %}
+<h1 style="margin: 0 0 10px 0">Manage event organizers</h1>
```
I would replace "manage" by "remove" since it's the only action you can do.

patjouk commented on templates/emails/new_user.html:
```diff
@@ -0,0 +1,25 @@
+<p>Here you can find instructions on editing website: http://organize.djangogirls.org/website/</p>
+
+<p>We also invited you to Django Girls Slack, where we chat, communicate and meet each other!</p>
+
```
This is absolutely awesome :heart_eyes: 
When I add someone to an existing event, I usually add info about the event email account: instructions on how to connect and that they should ask the main organizer to obtain the password.
It would be also cool to add the procedure for support ("You can contact hello@ if you need assistance but it’s usually faster to check our FAQ (link) or to post questions to our Google Group of Django Girls Organizers (link) or Slack channel (link) - there are more people that can help you! :)")

olasitarska replied:
Is there a standard response template that I can copy? Or maybe you can update those templates here and push to this branch?

patjouk commented:
This PR is awesome :heart: It will definitely make my job easier.
I have three more comments I couldn't inline:
- For the new_event command: I moved the setup template email to a canned answer in gmail: it would be great if it only shows the thing I need to copy past (email recipients, email subject, event's emails and website). Since I'm not the only one to deploy event, I would totally understand if you want to keep the full text.
- A confirmation before removing someone from the team would be great: we removed this feature because of the miss-clicks and I'm pretty sure it will happen again. How about a page saying "are you sure you want to remove X from your team" or a checkbox with a save button?
- You shouldn't be able to kick yourself from a team :wink:

patjouk commented on core/admin.py:
```diff
@@ -44,6 +49,83 @@ def get_readonly_fields(self, request, obj=None):
+            if form.is_valid():
+                user = form.save()
+                messages.success(request,
+                    '{} has been added to your event, yay!'.format(user.get_full_name()))
```
It should also say that the new organizer received an email with their credentials and was also added to Slack.

olasitarska commented:
@patjouk From your non-inline comment:

> - For the new_event command: I moved the setup template email to a canned answer in gmail: it would be great if it only shows the thing I need to copy past (email recipients, email subject, event's emails and website). Since I'm not the only one to deploy event, I would totally understand if you want to keep the full text.

Done!

> - A confirmation before removing someone from the team would be great: we removed this feature because of the miss-clicks and I'm pretty sure it will happen again. How about a page saying "are you sure you want to remove X from your team", a checkbox with a save button or an "undo" button?

Done as well :)

> - You shouldn't be able to kick yourself from a team :wink: 

This too!

> - Organizers added by other organizers are not added to Google Group. Is there a way to obtain a list every day or week of the email addresses that were added?

I added `date_joined` to be displayed in [this table](https://djangogirls.org/admin/core/user/), so you're able to sort by it and see who recently joined. Also added this to filtering, so you can filter on who joined last week, for example ;) Hope that's enough!
